File: ruby/lib/enolib/elements/section_element.rb
# frozen_string_literal: true

module Enolib
  class SectionElement < ElementBase
    attr_reader :instruction # TODO: Revisit this hacky exposition

    def _untouched
      unless instance_variable_defined?(:@yielded)
        return instance_variable_defined?(:@touched) ? false : @instruction
      end

      return @instruction if instance_variable_defined?(:@empty) && !@empty.instance_variable_defined?(:@touched)
      return @instruction if instance_variable_defined?(:@field) && !@field.instance_variable_defined?(:@touched)
      return @fieldset._untouched if instance_variable_defined?(:@fieldset)
      return @list._untouched if instance_variable_defined?(:@list)
      return @section._untouched if instance_variable_defined?(:@section)
    end

    def to_empty
      unless instance_variable_defined?(:@empty)
        unless @instruction[:type] == :empty
          # TODO: Below and in all implementations - why nil for key as second parameter?
          raise Errors::Validation.unexpected_element_type(@context, nil, @instruction, 'expected_empty')
        end

        @empty = Empty.new(@context, @instruction, @parent)
        @yielded = :empty
      end

      @empty
    end

    def to_field
      unless instance_variable_defined?(:@field)
        if instance_variable_defined?(:@yielded)
          raise TypeError, "This element was already yielded as #{PRETTY_TYPES[@yielded]} and can't be yielded again as a field."
        end

        unless @instruction[:type] == :field ||
               @instruction[:type] == :multiline_field_begin ||
               @instruction[:type] == :field_or_fieldset_or_list
          raise Errors::Validation.unexpected_element_type(@context, nil, @instruction, 'expected_field')
        end

        @field = Field.new(@context, @instruction, @parent)
        @yielded = :field
      end

      @field
    end

    def to_fieldset
      unless instance_variable_defined?(:@fieldset)
        if instance_variable_defined?(:@yielded)
          raise TypeError, "This element was already yielded as #{PRETTY_TYPES[@yielded]} and can't be yielded again as a fieldset."
        end

        unless @instruction[:type] == :fieldset || @instruction[:type] == :field_or_fieldset_or_list
          raise Errors::Validation.unexpected_element_type(@context, nil, @instruction, 'expected_fieldset')
        end

        @fieldset = Fieldset.new(@context, @instruction, @parent)
        @yielded = :fieldset
      end

      @fieldset
    end

    def to_list
      unless instance_variable_defined?(:@list)
        if instance_variable_defined?(:@yielded)
          raise TypeError, "This element was already yielded as #{PRETTY_TYPES[@yielded]} and can't be yielded again as a list."
        end

        unless @instruction[:type] == :list || @instruction[:type] == :field_or_fieldset_or_list
          raise Errors::Validation.unexpected_element_type(@context, nil, @instruction, 'expected_list')
        end

        @list = List.new(@context, @instruction, @parent)
        @yielded = :list
      end

      @list
    end

    def to_s
      "#<Enolib::SectionElement key=#{_key} yields=#{_yields}>"
    end

    def to_section
      unless instance_variable_defined?(:@section)
        unless @instruction[:type] == :section
          raise Errors::Validation.unexpected_element_type(@context, nil, @instruction, 'expected_section')
        end

        @section = Section.new(@context, @instruction, @parent)
        @yielded = :section
      end

      @section
    end

    def touch
      # TODO: Revisit setting touched on foreign instances
      if !instance_variable_defined?(:@yielded)
        @touched = true
      elsif instance_variable_defined?(:@empty)
        @empty.touched = true
      elsif instance_variable_defined?(:@field)
        @field.touched = true
      elsif instance_variable_defined?(:@fieldset)
        @fieldset.touch
      elsif instance_variable_defined?(:@list)
        @list.touch
      elsif instance_variable_defined?(:@section)
        @section.touch
      end
    end

    def yields_empty?
      @instruction[:type] == :empty
    end

    def yields_field?
      @instruction[:type] == :field ||
      @instruction[:type] == :multiline_field_begin ||
      @instruction[:type] == :field_or_fieldset_or_list
    end

    def yields_fieldset?
      @instruction[:type] == :fieldset ||
      @instruction[:type] == :field_or_fieldset_or_list
    end

    def yields_list?
      @instruction[:type] == :list ||
      @instruction[:type] == :field_or_fieldset_or_list
    end

    def yields_section?
      @instruction[:type] == :section
    end

    private

    def _yields
      if @instruction[:type] == :field_or_fieldset_or_list
          return "#{PRETTY_TYPES[:field]},#{PRETTY_TYPES[:fieldset]},#{PRETTY_TYPES[:list]}"
      end

      PRETTY_TYPES[@instruction[:type]]
    end
  end
end

File: python/enolib/elements/section_element.py
from ..errors.validation import Validation
from . import empty
from . import field
from . import fieldset
from . import list as list_module  # don't globally override built-in list function
from . import section
from .element_base import ElementBase
from ..constants import (
    EMPTY,
    FIELD,
    FIELDSET,
    FIELD_OR_FIELDSET_OR_LIST,
    LIST,
    MULTILINE_FIELD_BEGIN,
    PRETTY_TYPES,
    SECTION
)

class SectionElement(ElementBase):
    def __repr__(self):
        return f"<class SectionElement key={self._key()} yields={self._yields()}>"

    def _untouched(self):
        if not hasattr(self, '_yielded') and not hasattr(self, '_touched'):
            return self._instruction

        if hasattr(self, '_empty') and not hasattr(self._empty, '_touched'):
            return self._instruction

        if hasattr(self, '_field') and not hasattr(self._field, '_touched'):
            return self._instruction

        if hasattr(self, '_fieldset'):
            return self._fieldset._untouched()

        if hasattr(self, '_list'):
            return self._list._untouched()

        if hasattr(self, '_section'):
            return self._section._untouched()

    def _yields(self):
        if self._instruction['type'] == FIELD_OR_FIELDSET_OR_LIST:
            return f"{PRETTY_TYPES[FIELD]},{PRETTY_TYPES[FIELDSET]},{PRETTY_TYPES[LIST]}"

        return PRETTY_TYPES[self._instruction['type']]

    def to_empty(self):
        if not hasattr(self, '_empty'):
            if self._instruction['type'] != EMPTY:
                raise Validation.unexpected_element_type(self._context, None, self._instruction, 'expected_empty')

            self._empty = empty.Empty(self._context, self._instruction, self._parent)
            self._yielded = EMPTY

        return self._empty

    def to_field(self):
        if not hasattr(self, '_field'):
            if hasattr(self, '_yielded'):
                raise TypeError(f"This element was already yielded as {PRETTY_TYPES[self._yielded]} and can't be yielded again as a field.")

            if (self._instruction['type'] != FIELD and
                self._instruction['type'] != MULTILINE_FIELD_BEGIN and
                self._instruction['type'] != FIELD_OR_FIELDSET_OR_LIST):
                raise Validation.unexpected_element_type(self._context, None, self._instruction, 'expected_field')

            self._field = field.Field(self._context, self._instruction, self._parent)
            self._yielded = FIELD

        return self._field

    def to_fieldset(self):
        if not hasattr(self, '_fieldset'):
            if hasattr(self, '_yielded'):
                raise TypeError(f"This element was already yielded as {PRETTY_TYPES[self._yielded]} and can't be yielded again as a fieldset.")

            if self._instruction['type'] != FIELDSET and self._instruction['type'] != FIELD_OR_FIELDSET_OR_LIST:
                raise Validation.unexpected_element_type(self._context, None, self._instruction, 'expected_fieldset')

            self._fieldset = fieldset.Fieldset(self._context, self._instruction, self._parent)
            self._yielded = FIELDSET

        return self._fieldset

    def to_list(self):
        if not hasattr(self, '_list'):
            if hasattr(self, '_yielded'):
                raise TypeError(f"This element was already yielded as {PRETTY_TYPES[self._yielded]} and can't be yielded again as a list.")

            if self._instruction['type'] != LIST and self._instruction['type'] != FIELD_OR_FIELDSET_OR_LIST:
                raise Validation.unexpected_element_type(self._context, None, self._instruction, 'expected_list')

            self._list = list_module.List(self._context, self._instruction, self._parent)
            self._yielded = LIST

        return self._list

    def to_section(self):
        if not hasattr(self, '_section'):
            if self._instruction['type'] != SECTION:
                raise Validation.unexpected_element_type(self._context, None, self._instruction, 'expected_section')

            self._section = section.Section(self._context, self._instruction, self._parent)
            self._yielded = SECTION

        return self._section

    def touch(self):
        if not hasattr(self, '_yielded'):
            self._touched = True
        elif hasattr(self, '_empty'):
            self._empty._touched = True
        elif hasattr(self, '_field'):
            self._field._touched = True
        elif hasattr(self, '_fieldset'):
            self._fieldset.touch()
        elif hasattr(self, '_list'):
            self._list.touch()
        elif hasattr(self, '_section'):
            self._section.touch()

    def yields_empty(self):
        return self._instruction['type'] == EMPTY

    def yields_field(self):
        return (self._instruction['type'] == FIELD or
                self._instruction['type'] == MULTILINE_FIELD_BEGIN or
                self._instruction['type'] == FIELD_OR_FIELDSET_OR_LIST)

    def yields_fieldset(self):
        return (self._instruction['type'] == FIELDSET or
                self._instruction['type'] == FIELD_OR_FIELDSET_OR_LIST)

    def yields_list(self):
        return (self._instruction['type'] == LIST or
                self._instruction['type'] == FIELD_OR_FIELDSET_OR_LIST)

    def yields_section(self):
        return self._instruction['type'] == SECTION
